File: java-sdk/src/main/java/com/framework/protocol/adapter/ProtocolAdapter.java
package com.framework.protocol.adapter;

import com.framework.protocol.model.ExternalRequest;
import com.framework.protocol.model.ExternalResponse;
import com.framework.protocol.model.InternalRequest;
import com.framework.protocol.model.InternalResponse;

import java.util.List;

/**
 * 协议适配器接口
 * 
 * 负责外部协议和内部协议之间的转换，保持消息语义一致性。
 * 
 * 需求: 4.1, 4.2, 4.5
 */
public interface ProtocolAdapter {
    
    /**
     * 外部协议转内部协议
     * 
     * @param externalRequest 外部请求
     * @return 内部请求
     */
    InternalRequest transformRequest(ExternalRequest externalRequest);
    
    /**
     * 内部协议转外部协议
     * 
     * @param internalResponse 内部响应
     * @param originalRequest  原始外部请求（用于保持协议上下文）
     * @return 外部响应
     */
    ExternalResponse transformResponse(InternalResponse internalResponse, ExternalRequest originalRequest);
    
    /**
     * 内部协议转外部协议（无原始请求上下文时使用）
     * 
     * @param internalResponse 内部响应
     * @return 外部响应
     */
    ExternalResponse transformResponse(InternalResponse internalResponse);
    
    /**
     * 获取支持的外部协议类型列表
     * 
     * @return 协议类型列表
     */
    List<String> getSupportedProtocols();
    
    /**
     * 判断是否支持指定的外部协议
     * 
     * @param protocol 协议类型
     * @return 是否支持
     */
    boolean supportsProtocol(String protocol);
}

File: php-sdk/src/Protocol/Adapter/ProtocolAdapter.php
<?php

declare(strict_types=1);

namespace Framework\Protocol\Adapter;

/**
 * 协议适配器 - 外部协议与内部协议转换
 */
class ProtocolAdapter
{
    /**
     * 外部请求转内部格式
     */
    public function transformRequest(array $externalRequest): array
    {
        return [
            'service'   => $externalRequest['service'] ?? $this->extractService($externalRequest),
            'method'    => $externalRequest['method'] ?? $this->extractMethod($externalRequest),
            'payload'   => $externalRequest['body'] ?? $externalRequest['params'] ?? [],
            'headers'   => $externalRequest['headers'] ?? [],
            'traceId'   => $externalRequest['traceId'] ?? uniqid('trace-', true),
            'timestamp' => time(),
            'protocol'  => $externalRequest['protocol'] ?? 'unknown',
        ];
    }

    /**
     * 内部响应转外部格式
     */
    public function transformResponse(array $internalResponse, string $protocol = 'rest'): array
    {
        return match ($protocol) {
            'jsonrpc' => [
                'jsonrpc' => '2.0',
                'result'  => $internalResponse['data'] ?? $internalResponse,
                'id'      => $internalResponse['id'] ?? null,
            ],
            'mqtt' => [
                'topic'   => $internalResponse['topic'] ?? 'response',
                'payload' => json_encode($internalResponse['data'] ?? $internalResponse),
            ],
            default => [
                'status'  => $internalResponse['status'] ?? 200,
                'data'    => $internalResponse['data'] ?? $internalResponse,
                'traceId' => $internalResponse['traceId'] ?? null,
            ],
        };
    }

    public function getSupportedProtocols(): array
    {
        return ['rest', 'websocket', 'jsonrpc', 'mqtt'];
    }

    private function extractService(array $request): string
    {
        // 从URL路径提取服务名: /service/method -> service
        $path = $request['path'] ?? '';
        $parts = explode('/', trim($path, '/'));
        return $parts[0] ?? 'unknown';
    }

    private function extractMethod(array $request): string
    {
        $path = $request['path'] ?? '';
        $parts = explode('/', trim($path, '/'));
        return $parts[1] ?? 'call';
    }
}
